### PR TITLE
Feature/mds3337 admin control

### PR DESCRIPTION
**Pull request by mmumby**
# Main
- Added the ability to Get, Create, Edit, Delete, Re-order standard permit conditions on the admin tab
- Updated the admin tab navigation to include permission-based routes

# New Admin Permissions
 - Navigating to the condition tab without "core_edit_template_conditions" permissions (should never get to this point as the user will also be missing the "view_admin_route" permissions)
Screen Shot 2021-08-05 at 1 32 44 PM
- Navigating to the Admin tab without "view_admin_route" permissions (still have admin permissions to override the admin button logic)
Screen Shot 2021-08-05 at 1 30 21 PM
 - Navigating to the condition tab without "core_edit_template_conditions" permissions
 
 On the admin tab as a "senior permitting inspector" - only has access to the conditions tab, nothing else
Screen Shot 2021-08-05 at 1 35 31 PM
On the admin tab as a core admin has access to everything
Screen Shot 2021-08-05 at 1 37 20 PM

### CODE REVIEW

File: openshift4/templates/filesystem-provider.dc.yaml [hunk 1]
```diff
@@ -2,6 +2,12 @@ kind: Template
 apiVersion: v1
 metadata:
   name: mds-filesystem-provider-dc
+parameters:
+  - name: APPNAME
+    displayName: AppName
+    description: Application name for grouping pods in Openshift
+    value: Core
+    required: true
 objects:
```
Comment (mmumby) [inline openshift4/templates/filesystem-provider.dc.yaml]: @jon-funk I'm not sure why these are in our PR - I'm thinking I didn't resolve the merge conflicts correctly. Can you let me know what this and the redis.dc.yaml files should be.

Comment (mmumby) [reply openshift4/templates/filesystem-provider.dc.yaml]: @yasserhu do you mind taking a look at these two yaml files and advise. I think this is the result of merge conflicts as we did not add this code.

File: openshift4/templates/redis.dc.yaml [hunk 2]
```diff
@@ -14,6 +14,12 @@ metadata:
     openshift.io/provider-display-name: Red Hat, Inc.
     openshift.io/documentation-url: https://github.com/sclorg/redis-container/tree/master/3.2
     openshift.io/support-url: https://access.redhat.com
+parameters:
```
Comment (mmumby) [inline openshift4/templates/redis.dc.yaml]: @yasserhu should this one be here too? I can copy the file from dev and remove it if not

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py [hunk 3]
```diff
@@ -1,19 +1,18 @@
...
+from app.extensions import db
 from app.api.utils.models_mixins import SoftDeleteMixin, AuditMixin, Base
+from app.api.utils.list_lettering_helpers import num_to_letter, num_to_roman
 
 
 class StandardPermitConditions(SoftDeleteMixin, AuditMixin, Base):
     __tablename__ = 'standard_permit_conditions'
 
-    standard_permit_condition_id = db.Column(db.Integer, primary_key=True)
+    standard_permit_condition_id = db.Column(db.Integer, primary_key=True,server_default=FetchedValue())
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py]: Looks like the formatter needs to run in this file. There should be a space after the ",".

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py [hunk 4]
```diff
@@ -27,11 +26,47 @@ class StandardPermitConditions(SoftDeleteMixin, AuditMixin, Base):
...
     display_order = db.Column(db.Integer, nullable=False)
-    sub_conditions = db.relationship(
+    all_sub_conditions = db.relationship(
         'StandardPermitConditions',
         lazy='joined',
+        order_by='asc(StandardPermitConditions.display_order)',
         backref=backref('parent', remote_side=[standard_permit_condition_id]))
 
+    def __repr__(self):
+        return '<StandardPermitConditions %r, %r>' % (self.standard_permit_condition_id,
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py]: There is a way to get the class name automatically. Can you adjust it to do that? Look at other "repr" functions in other classes for an example.

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py [hunk 5]
```diff
@@ -27,11 +26,47 @@ class StandardPermitConditions(SoftDeleteMixin, AuditMixin, Base):
...
+                                                      self.standard_permit_condition_guid)
+
+    def __str__(self):
+        return f'<StandardPermitConditions> standard_permit_condition_id: {self.standard_permit_condition_id}, standard_permit_condition_guid: {self.standard_permit_condition_guid}, condition: {self.condition}, condition_category_code: {self.condition_category_code}, condition_type_code: {self.condition_type_code}, notice_of_work_type: {self.notice_of_work_type}, parent_standard_permit_condition_id: {self.parent_standard_permit_condition_id}, display_order: {self.display_order}, all_sub_conditions: {self.all_sub_conditions}'
+
+    @hybrid_property
+    def sub_conditions(self):
+        return [x for x in self.all_sub_conditions if x.deleted_ind == False]
+
+    # The FE is expecting the same keys are permitConditions
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py]: Can we remove this comment? I don't understand it and am not sure if it's helpful. 😜

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py [hunk 6]
```diff
@@ -27,11 +26,47 @@ class StandardPermitConditions(SoftDeleteMixin, AuditMixin, Base):
...
+    @hybrid_property
+    def permit_condition_id(self):
+        return self.standard_permit_condition_id
+
+    @hybrid_property
+    def parent_permit_condition_id(self):
+        return self.parent_standard_permit_condition_id
+
+    @hybrid_property
+    def step(self):
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py:56]: Cool function!

File: services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_list_resource.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+class StandardPermitConditionsListResource(Resource, UserMixin):
+    @api.doc(description='Get all standard permit conditions for a notice of work type')
+    @requires_role_edit_standard_permit_conditions
+    @api.marshal_with(STANDARD_PERMIT_CONDITION_MODEL, code=200, envelope='records')
+    def get(self, notice_of_work_type):
+        standard_permit_condition = StandardPermitConditions.find_by_notice_of_work_type_code(
+            notice_of_work_type)
+        if not standard_permit_condition:
+            raise BadRequest(
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_list_resource.py]: Can you return a NotFound here?

File: services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_resource.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            old_display_order = old_condition.display_order
+
+        try:
+            condition = StandardPermitConditions._schema().load(
+                request.json,
+                instance=StandardPermitConditions.find_by_standard_permit_condition_guid(standard_permit_condition_guid))
+        except MarshmallowError as e:
+            raise BadRequest(e)
+
+        conditions=[]
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_resource.py]: I think the formatter needs to run in this file. There should be spaces around the equals sign. I recommend enabling formatting on save.

File: services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_resource.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+            for i, cond in enumerate(conditions):
+                cond.display_order = i + 1
+                cond.save(commit=False)
+
+        db.session.commit()
+        return condition
+
+
+    @api.doc(description='delete a standard permit condition')
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_resource.py]: We should use consistent casing and punctuation for our API documentation

File: services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_resource.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        if conditions:
+            for i, condition in enumerate(sorted(conditions, key=lambda x: x.display_order)):
+                condition.display_order = i + 1
+                condition.save(commit=False)
+
+        db.session.commit()
+
+        return ('', 204)
+
+
```
Comment (exzizt) [inline services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_resource.py]: I think the formatter should remove spaces at the bottom of the file? I think ideally we just have 1 space.

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py (hunk 3 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py]: You are right :) , I applied the formatter.

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py (hunk 4 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py]: you are right,  it is better.

File: services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py (hunk 5 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/permits/permit_conditions/models/standard_permit_conditions.py]: Removed.

File: services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_list_resource.py (hunk 7 above)
Comment (areyeslo) [reply services/core-api/app/api/mines/permits/permit_conditions/resources/standard_permit_conditions_list_resource.py]: Absolutely, makes more sense

**Review by exzizt — APPROVED** [review]
Amazing work you guys! Great teamwork on this one 💯 😃

**Pull request merged**